### ISSUE TITLE
Class imbalance issues

### ISSUE THREAD

**Issue by RuthKassahun** [ultralytics/ultralytics#20921]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi everyone! I’m running into an issue with class imbalance in my dataset. My first class has around 15K bounding boxes but my second and third classes are much smaller (1.4K and 600). I worked with a similar imbalanced dataset before and the network worked fairly well after I gave higher class weights for under represented classes
```
        class_weights = torch.tensor([0.0,0.3,0.7]).cuda()
        loss = F.cross_entropy(preds, batch["cls"], weight=class_weights)
```

This time around it's performing very poorly. What are the best work around in this situation? Thanks!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @RuthKassahun, thank you for reaching out and sharing your experience with Ultralytics 🚀! Your question about handling class imbalance is a common one, and we appreciate the detailed context you’ve provided.

If this is a 🐛 Bug Report, please help us by providing a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) so we can investigate further.

If this is a custom training ❓ Question, please include as much information as possible, such as sample images from your dataset, your training logs, and confirm you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

For great resources on using Ultralytics models, check out our [Docs](https://docs.ultralytics.com/), including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage guides. Many common training and class imbalance questions are addressed there.

Join our community for more help and discussion:
- For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth topics, use [Discourse](https://community.ultralytics.com/)
- Or join our [Subreddit](https://reddit.com/r/Ultralytics) for community support

## Upgrade

Please ensure you’re using the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these verified environments (dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response. An Ultralytics engineer will review your issue and provide further assistance soon. Thank you for your patience and for contributing to the Ultralytics community!

**Comment by Y-T-G** [comment]
The code you highlighted isn't used for detection training. It's for training classification models. So the changes you made there never had any effect.

**Comment by Y-T-G** [comment]
If you want to adjust weights you can use this

https://github.com/ultralytics/ultralytics/issues/15044#issuecomment-2276387717

or you can try this

https://y-t-g.github.io/tutorials/yolo-class-balancing/

But neither of them compensates for lack of sufficient data. Your dataset is not just imbalanced, there's also not enough data for the minority classes.

https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/#dataset

**Comment by RuthKassahun** [comment]
> If you want to adjust weights you can use this
> 
> [#15044 (comment)](https://github.com/ultralytics/ultralytics/issues/15044#issuecomment-2276387717)
> 
> or you can try this
> 
> https://y-t-g.github.io/tutorials/yolo-class-balancing/
> 
> But neither of them compensates for lack of sufficient data. Your dataset is not just imbalanced, there's also not enough data for the minority classes.
> 
> https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/#dataset

Thank you for sharing the resources. I followed your tutorial and trained using the weighted data loader. I noticed the weight didn't balance as I expected. Is there a way to suppress the high class and only focus on the low classes?

![Image](https://github.com/user-attachments/assets/bbfe228c-a55e-4603-85c1-cfacbfeacc96)

**Comment by Y-T-G** [comment]
It's probably because your classes are coupled. So they can't be balanced by increasing/decreasing images which is what the weighted dataloader does. You can only use class weights in losses.

**Comment by Y-T-G** [comment]
Duplicate of https://github.com/ultralytics/ultralytics/issues/21082
